src/main/java/soot/javaToJimple/LocalGenerator.java

```diff
@@ -23,203 +23,158 @@
  */
 
 import java.util.HashSet;
-import java.util.Set;
+
 import soot.Body;
+import soot.BooleanType;
+import soot.ByteType;
+import soot.CharType;
+import soot.DoubleType;
+import soot.FloatType;
+import soot.IntType;
 import soot.Local;
+import soot.LongType;
+import soot.RefLikeType;
+import soot.ShortType;
+import soot.Type;
+import soot.UnknownType;
+import soot.VoidType;
+import soot.jimple.Jimple;
 import soot.jimple.toolkits.typing.fast.Integer127Type;
 import soot.jimple.toolkits.typing.fast.Integer1Type;
 import soot.jimple.toolkits.typing.fast.Integer32767Type;
+import soot.util.Chain;
 
 public class LocalGenerator {
 
-  protected final soot.Body body;
-
-  public LocalGenerator(Body b) {
-    body = b;
-  }
-
-  protected transient Set<String> localNames = null;
+  protected final Chain<Local> locals;
 
-  protected boolean bodyContainsLocal(String name) {
-    return localNames.contains(name);
-  }
+  private int tempInt = -1;
+  private int tempVoid = -1;
+  private int tempBoolean = -1;
+  private int tempLong = -1;
+  private int tempDouble = -1;
+  private int tempFloat = -1;
+  private int tempRefLikeType = -1;
+  private int tempByte = -1;
+  private int tempShort = -1;
+  private int tempChar = -1;
+  private int tempUnknownType = -1;
 
-  private void initLocalNames() {
-    localNames = new HashSet<String>();
-    for (Local l : body.getLocals()) {
-      localNames.add(l.getName());
-    }
+  public LocalGenerator(Body b) {
+    this.locals = b.getLocals();
   }
 
   /** generates a new soot local given the type */
-  public soot.Local generateLocal(soot.Type type) {
+  public Local generateLocal(Type type) {
 
     // store local names for enhanced performance
-    initLocalNames();
+    HashSet<String> localNames = new HashSet<String>();
```

Comment:
Why did you move the local name cache inside the method, instead of keeping it as a class member? When generating many locals, this might impact performance.

Reply:
The previous version actually had the line ```localNames = null;``` at the end of the ```generateLocal(Type)``` method so the set was rebuilt at for each call to the method anyway. Thus, logically equivalent to my updated code. I simply removed the field since it was essentially useless.

I would guess it was rebuilt for each call to the method because the set can easily become out of sync with the Chain of locals. I do have an approach that can address this issue and achieve the performance from caching that you are wanting: go back to storing the Set in a field and add an additional field to store the expected modification count of the Chain; instead of setting the Set to ```null``` at the end of each call to the method the start of the method should compare the actual modification count of the chain to the expected modification count and only rebuild if necessary. I can update the PR to use this improved approach.

Reply:
Your approach sounds good, that should improve performance quite a bit. I agree, the previous version did not have a cache, and that was actually bad right from the start. Could you update your PR?

Reply:
I will make the changes mentioned and update the PR.